File: 02-01-2026/anagram_groups.js
function groupAnagrams(words) {
  const anagramGroups = {};

  for (const word of words) {
    const sortedWord = word.split("").sort().join("");

    if (!anagramGroups[sortedWord]) anagramGroups[sortedWord] = [];
    anagramGroups[sortedWord].push(word);
  }

  return Object.values(anagramGroups);
}

console.log(groupAnagrams(["listen", "silent", "hello", "enlist", "world"])); // [["listen", "silent", "enlist"], ["hello"], ["world"]].
console.log(groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"])); // [["ate", "eat", "tea"], ["bat"], ["nat", "tan"]].
console.log(
  groupAnagrams([
    "care",
    "race",
    "acre",
    "pots",
    "stop",
    "tops",
    "opts",
    "post",
    "spot",
    "evil",
    "vile",
    "live",
    "veil",
  ]),
); // [["acre", "care", "race"], ["evil", "live", "veil", "vile"], ["opts", "post", "pots", "spot", "stop", "tops"]].
console.log(
  groupAnagrams([
    "algorithms",
    "logarithms",
    "education",
    "cautioned",
    "auctioned",
    "triangle",
    "integral",
    "alerting",
    "relating",
  ]),
); // [["alerting", "integral", "relating", "triangle"], ["algorithms", "logarithms"], ["auctioned", "cautioned", "education"]].

File: 02-05-2026/anagram_groups.py
def group_anagrams(words):
    anagram_dict = {}
    
    for word in words:
        sorted_word = ''.join(sorted(word))
        if sorted_word in anagram_dict:
            anagram_dict[sorted_word].append(word)
        else:
            anagram_dict[sorted_word] = [word]

    return list(anagram_dict.values())

print(group_anagrams(["listen", "silent", "hello", "enlist", "world"])) # [["listen", "silent", "enlist"], ["hello"], ["world"]].
print(group_anagrams(["eat", "tea", "tan", "ate", "nat", "bat"])) # [["ate", "eat", "tea"], ["bat"], ["nat", "tan"]].
print(group_anagrams(["care", "race", "acre", "pots", "stop", "tops", "opts", "post", "spot", "evil", "vile", "live", "veil"])) # [["acre", "care", "race"], ["evil", "live", "veil", "vile"], ["opts", "post", "pots", "spot", "stop", "tops"]].
print(group_anagrams(["algorithms", "logarithms", "education", "cautioned", "auctioned", "triangle", "integral", "alerting", "relating"])) # [["alerting", "integral", "relating", "triangle"], ["algorithms", "logarithms"], ["auctioned", "cautioned", "education"]].